src/operator/random/sample_multinomial_op.h

```diff
@@ -155,9 +158,11 @@ void SampleMultinomialForward(const nnvm::NodeAttrs& attrs,
     Tensor<xpu, 1, float> uniform =
       ctx.requested[1].get_space_typed<xpu, 1, float>(Shape1(N*M), s);
     prnd->SampleUniform(&uniform, 0, 1);
-    Kernel<SampleMultinomialKernel, xpu>::Launch(
-      s, N, K, M, inputs[0].dptr<DType>(), uniform.dptr_, outputs[0].dptr<int>(),
-      param.get_prob ? outputs[1].dptr<DType>() : nullptr);
+    MSHADOW_TYPE_SWITCH(outputs[0].type_flag_, IType, {
```

Comment:
This kind of 2 layer switches is very slow to compile. Why do we need type support for output?

Reply:
Sometimes the multinomial samples need further processing in floating point arithmetic, so the samples need to be copied into a new array of floating point type which slow down the training. For example, in RBM, the samples need to be applied by `linalg.gemm` which supports only floating point arrays.

Reply:
A simple cast shouldn't cost that much?

This kind of nested switches are really slow to compile and makes the binary much bigger.
We need to make sure it really justifies the cost

Reply:
The binary size increases about 0.1% for both shared and static library (CUDA, CUDNN, MKL). Compiling mxnet already takes quite long time, so the relative increase of the compile time is also tiny.

I'm working with some variants of RBM and the use of `.astype('float32')` in several places increases training time over 20%. In the case of usual basic RBM, it increases about 10% of training time in my [test](https://github.com/asitstands/incubator-mxnet/blob/e96c79d84e00e5b7d26b18c8faba0daaac812f5e/example/restricted-boltzmann-machine/binary_rbm.py#L86) for mnist. Of course, it depends on the hyperparameters and data. However, I think that, in general, the cost cannot be ignored for applications using heavy Monte Carlo samplings of discrete states.